File: backend/app/services/master_data_service.py
"""
Master Data Service
"""

from typing import List
from sqlalchemy.orm import Session
from app.repositories.master_data_repository import (
    ComplaintRepository,
    DiagnosisRepository,
    LabTestRepository,
    MedicineRepository,
)
from app.models.master_data import ComplaintMaster, DiagnosisMaster, LabTestMaster
from app.models.medicine import MedicineMaster


class MasterDataService:
    def __init__(self, db: Session):
        self.db = db
        self.complaint_repo = ComplaintRepository(db)
        self.diagnosis_repo = DiagnosisRepository(db)
        self.labtest_repo = LabTestRepository(db)
        self.medicine_repo = MedicineRepository(db)

    def search_complaints(self, query: str, limit: int = 20) -> List[ComplaintMaster]:
        if not query:
            return []
        return self.complaint_repo.search(query, limit)

    def search_diagnosis(self, query: str, limit: int = 20) -> List[DiagnosisMaster]:
        if not query:
            return []
        return self.diagnosis_repo.search(query, limit)

    def search_lab_tests(self, query: str, limit: int = 20) -> List[LabTestMaster]:
        if not query:
            return []
        return self.labtest_repo.search(query, limit)

    def search_medicines(self, query: str, limit: int = 20) -> List[MedicineMaster]:
        """Search medicines by name, generic name, code, or category."""
        if not query:
            return []
        return self.medicine_repo.search(query, limit)

File: opd_web/src/services/masterDataService.ts
import { api } from './api';

export interface MasterDataItem {
    id: string;
    code: string;
    category?: string;
}

export interface Complaint extends MasterDataItem {
    complaint: string;
}

export interface Diagnosis extends MasterDataItem {
    diagnosis: string;
}

export interface LabTest extends MasterDataItem {
    test_name: string;
}

export interface Medicine extends MasterDataItem {
    name: string;
    generic_name?: string;
    specialization: string;
    dosage_form?: string;
    strength?: string;
}

class MasterDataService {
    async searchComplaints(query: string, limit: number = 20): Promise<Complaint[]> {
        if (!query || query.length < 1) return [];
        return api.get<Complaint[]>(`/master/complaints/search?q=${encodeURIComponent(query)}&limit=${limit}`);
    }

    async searchDiagnosis(query: string, limit: number = 20): Promise<Diagnosis[]> {
        if (!query || query.length < 1) return [];
        return api.get<Diagnosis[]>(`/master/diagnosis/search?q=${encodeURIComponent(query)}&limit=${limit}`);
    }

    async searchLabTests(query: string, limit: number = 20): Promise<LabTest[]> {
        if (!query || query.length < 1) return [];
        return api.get<LabTest[]>(`/master/lab-tests/search?q=${encodeURIComponent(query)}&limit=${limit}`);
    }

    async searchMedicines(query: string, limit: number = 20): Promise<Medicine[]> {
        if (!query || query.length < 3) return [];
        return api.get<Medicine[]>(`/master/medicines/search?q=${encodeURIComponent(query)}&limit=${limit}`);
    }
}

export const masterDataService = new MasterDataService();
export default masterDataService;
